changefeedccl: fix bug with initial scans after restarts

ajwerner opened the pull request:
In the recent PR which introduced the option to opt in to the initial scan
with a cursor, a bug was introduced whereby initial scans would be performed
additionally when changefeeds were resumed and that option was specified.

The reason this happened due to my lack of clarity as to how the
`initialHighWater` and `statementTime` were being communicated to the
changeAggregator. I've now lifted the interpretation of the fields up above
the changeAggregator constructor.

Release justification: bug fixes and low-risk updates to new functionality.

Release note: None

Commit:
changefeedccl: fix bug with initial scans after restarts

In the recent PR which introduced the option to opt in to the initial scan
with a cursor, a bug was introduced whereby initial scans would be performed
additionally when changefeeds were resumed and that option was specified.

The reason this happened due to my lack of clarity as to how the
`initialHighWater` and `statementTime` were being communicated to the
changeAggregator. I've now lifted the interpretation of the fields up above
the changeAggregator constructor.

Release justification: bug fixes and low-risk updates to new functionality

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45821)
<!-- Reviewable:end -->

Commit:
changefeedccl: cosmetic changes to changeAggregator.Start

I was cleaning up comments and just started moving stuff around. It could go
further but I didn't want to throw this away.

Release note: None

Release justification: bug fixes and low-risk updates to new functionality.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I think @danhhz should also take a look at this when he's back tomorrow because it touches code he wrote and addresses a few of his PRs.

> Reviewed 5 of 5 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45821#-:-M1zsW_EAc6P0qntDZ7j:b71s9w4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/ccl/changefeedccl/changefeed_dist.go, line 79 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45821#-M1zqpkb0yPlBXCV-8KC:-M1zqpkcBEw974q1-G7_:bsliqv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa719d354dde8e3f857f278219c4679f8f2a562c/pkg/ccl/changefeedccl/changefeed_dist.go#L79)):*
> ```Go
> 	// NB: A non-empty high water indicates that we have checkpointed a resolved
> 	// timestamp. Skipping the initial scan is equivalent to starting the
> 	// changefeed from a checkpoint at its start time. Initial the progress based
> ```

"Initially" or "Initialize"?
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45821#-M1zrZz4D_fghAHWx-9r:-M1zrZz4D_fghAHWx-9s:bzc76ht) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa719d354dde8e3f857f278219c4679f8f2a562c/pkg/ccl/changefeedccl/changefeed_processors.go#L261)):*
> ```Go
> // watches, if we don't then we don't. If we have a cursor, we'll set the
> // statement time to be equal to the cursor timestamp. This loop over watches
> // is exists for a feature which is no longer in use - the checkpointing of
> ```

`s/is //`
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45821#-M1zrdfoEACkqhj20C2t:-M1zrdfoEACkqhj20C2u:be592zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa719d354dde8e3f857f278219c4679f8f2a562c/pkg/ccl/changefeedccl/changefeed_processors.go#L261)):*
> ```
> for a feature which is no longer in use
> ```

Should we get rid of it then?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
TFTR! I'll wait for @danhhz 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45821#-:-M2-W3Rt5DUeF70eB2eF:bmur02z)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/ccl/changefeedccl/changefeed_dist.go, line 79 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45821#-M1zqpkb0yPlBXCV-8KC:-M2-UQy_A_UMjrxnEccA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa719d354dde8e3f857f278219c4679f8f2a562c/pkg/ccl/changefeedccl/changefeed_dist.go#L79)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"Initially" or "Initialize"?
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45821#-M1zrZz4D_fghAHWx-9r:-M2-USb25g4K66MiKoL-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa719d354dde8e3f857f278219c4679f8f2a562c/pkg/ccl/changefeedccl/changefeed_processors.go#L261)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/is //`
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45821#-M1zrdfoEACkqhj20C2t:-M2-UTLK8NMwian2fX3q:bn5y1sf) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa719d354dde8e3f857f278219c4679f8f2a562c/pkg/ccl/changefeedccl/changefeed_processors.go#L261)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> for a feature which is no longer in use
> ```

Should we get rid of it then?
</blockquote></details>

I added a TODO and some commentary on a related issue. I think it's valuable in some cases but it needs to be more considered in its use. 


<!-- Sent from Reviewable.io -->

danhhz commented:
I didn't look closely but approach seems good

ajwerner commented:
TFTR!

bors r+

ajwerner commented:
bors r-

I think this needs a justification

ajwerner commented:
bors r+

ajwerner commented:
bors r+
